expression/builtin_cast.go

```diff
@@ -1400,10 +1401,9 @@ func WrapWithCastAsJSON(expr Expression, ctx context.Context) Expression {
 		return expr
 	}
 	tp := &types.FieldType{
-		Tp:   mysql.TypeJSON,
-		Flen: 12582912,
-		// Here we set decimal to -1 to indicate this is a post-wrapped cast.
-		Decimal: -1,
+		Tp:      mysql.TypeJSON,
+		Flen:    12582912,                    // FIXME: Here the Flen is not trusted.
+		Decimal: decimal4CastJSONPostWrapped, // Here we do *COERCE*, not *CAST*.
```

Comment:
will it better to use an intern flag which is mentioned [here](https://github.com/pingcap/tidb/pull/4379#discussion_r136731267)